// @typegoose/typegoose/lib/internal/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const mongoose = require("mongoose");
const util_1 = require("util");
const constants_1 = require("./constants");
const data_1 = require("./data");
const errors_1 = require("./errors");
const primitives = ['String', 'Number', 'Boolean', 'Date', 'Decimal128', 'ObjectID', 'Array'];
/**
 * Returns true, if it includes the Type
 * @param Type The Type
 * @returns true, if it includes it
 */
function isPrimitive(Type) {
    return primitives.includes(Type.name);
}
exports.isPrimitive = isPrimitive;
/**
 * Returns true, if it is an Object
 * @param Type The Type
 * @returns true, if it is an Object
 */
function isObject(Type) {
    let prototype = Type.prototype;
    let name = Type.name;
    while (name) {
        if (name === 'Object') {
            return true;
        }
        prototype = Object.getPrototypeOf(prototype);
        name = prototype ? prototype.constructor.name : null;
    }
    return false; // can this even return false?
    // return !isNullOrUndefined(Type) && (typeof Type === 'object' || Type.name === 'Object') && !Array.isArray(Type);
}
exports.isObject = isObject;
/**
 * Returns true, if it is an Number
 * @param Type The Type
 * @returns true, if it is an Number
 */
function isNumber(Type) {
    return Type.name === 'Number';
}
exports.isNumber = isNumber;
/**
 * Returns true, if it is an String
 * @param Type The Type
 * @returns true, if it is an String
 */
function isString(Type) {
    return Type.name === 'String';
}
exports.isString = isString;
/**
 * Initialize as Object
 * @param name The Name of the Schema
 * @param key The Property key to set
 */
function initAsObject(name, key) {
    if (!data_1.schemas.get(name)) {
        data_1.schemas.set(name, {});
    }
    data_1.schemas.get(name)[key] = {};
}
exports.initAsObject = initAsObject;
/**
 * Initialize as Array
 * @param name The Name of the Schema
 * @param key The Property key to set
 */
function initAsArray(name, key) {
    if (!data_1.schemas.get(name)) {
        data_1.schemas.set(name, {});
    }
    data_1.schemas.get(name)[key] = [{}];
}
exports.initAsArray = initAsArray;
/**
 * Get the Class for a given Document
 * @param document The Document
 */
function getClassForDocument(document) {
    const modelName = document.constructor.modelName;
    return data_1.constructors.get(modelName);
}
exports.getClassForDocument = getClassForDocument;
/**
 * Return true if there are Options
 * @param options The raw Options
 */
function isWithStringValidate(options) {
    return !util_1.isNullOrUndefined(options.match
        || options.enum
        || options.minlength
        || options.maxlength);
}
exports.isWithStringValidate = isWithStringValidate;
/**
 * Return true if there are Options
 * @param options The raw Options
 */
function isWithStringTransform(options) {
    return !util_1.isNullOrUndefined(options.lowercase || options.uppercase || options.trim);
}
exports.isWithStringTransform = isWithStringTransform;
/**
 * Return true if there are Options
 * @param options The raw Options
 */
function isWithNumberValidate(options) {
    return !util_1.isNullOrUndefined(options.min || options.max);
}
exports.isWithNumberValidate = isWithNumberValidate;
const virtualOptions = ['localField', 'foreignField'];
/**
 * Check if Options include Virtual Populate Options
 * @param options RawOptions of the Prop
 */
function isWithVirtualPOP(options) {
    return Object.keys(options).some((v) => virtualOptions.includes(v));
}
exports.isWithVirtualPOP = isWithVirtualPOP;
exports.allVirtualoptions = virtualOptions.slice(0);
exports.allVirtualoptions.push('ref');
/**
 * Check if All the required Options are present
 * @param options RawOptions of the Prop
 */
function includesAllVirtualPOP(options) {
    return exports.allVirtualoptions.every((v) => Object.keys(options).includes(v));
}
exports.includesAllVirtualPOP = includesAllVirtualPOP;
/**
 * Check if the given value has options of "IModelOptions"
 * @param value The Value to evaulate
 * @internal
 */
function isModelOptions(value) {
    return value && (typeof value.schemaOptions === 'object' ||
        typeof value.options === 'object');
}
/**
 * Merge value & existing Metadata & Save it to the class
 * Difference with "mergeMetadata" is that this one DOES save it to the class
 * @param key Metadata key
 * @param value Raw value
 * @param cl The constructor
 * @internal
 */
function assignMetadata(key, value, cl) {
    if (util_1.isNullOrUndefined(value)) {
        return value;
    }
    const newValue = mergeMetadata(key, value, cl);
    Reflect.defineMetadata(key, newValue, cl);
    return newValue;
}
exports.assignMetadata = assignMetadata;
/**
 * Merge value & existing Metadata
 * Difference with "assignMetadata" is that this one DOES NOT save it to the class
 * @param key Metadata key
 * @param value Raw value
 * @param cl The constructor
 * @internal
 */
function mergeMetadata(key, value, cl) {
    if (typeof key !== 'string') {
        throw new TypeError(`"${key}"(key) is not a string! (assignMetadata)`);
    }
    if (typeof cl !== 'function') {
        throw new errors_1.NoValidClass(cl);
    }
    const current = Object.assign({}, Reflect.getMetadata(key, cl) || {});
    if (util_1.isNullOrUndefined(value)) {
        return current;
    }
    // the following checks are needed, so that the new value dosnt override the full options
    // "deepmerge" cannot be used because of the other options like "existingMongoose"
    if (isModelOptions(value) && !util_1.isNullOrUndefined(current.schemaOptions)) {
        value.schemaOptions = Object.assign(current.schemaOptions, value.schemaOptions);
    }
    if (isModelOptions(value) && !util_1.isNullOrUndefined(current.options)) {
        value.options = Object.assign(current.options, value.options);
    }
    return Object.assign(current, value);
}
exports.mergeMetadata = mergeMetadata;
/**
 * Merge only schemaOptions from ModelOptions of the class
 * @param value The value to use
 * @param cl The Class to get the values from
 */
function mergeSchemaOptions(value, cl) {
    if (typeof cl !== 'function') {
        throw new errors_1.NoValidClass(cl);
    }
    const current = Reflect.getMetadata(constants_1.DecoratorKeys.ModelOptions, cl) || {};
    const evaluated = current && current.schemaOptions ? current.schemaOptions : {};
    if (util_1.isNullOrUndefined(value)) {
        return evaluated;
    }
    return Object.assign(evaluated, value || {});
}
exports.mergeSchemaOptions = mergeSchemaOptions;
/**
 * Get the correct name of the class's model
 * (with suffix)
 * @param cl The Class
 */
function getName(cl) {
    const options = Reflect.getMetadata(constants_1.DecoratorKeys.ModelOptions, cl) || {};
    const baseName = cl.name;
    if (options.options && options.options.automaticName) {
        const suffix = (options.options ? options.options.customName : undefined) ||
            (options.schemaOptions ? options.schemaOptions.collection : undefined);
        return suffix ? `${baseName}_${suffix}` : baseName;
    }
    if (options.options && typeof options.options.customName === 'string') {
        if (options.options.customName.length <= 0) {
            throw new TypeError(`"customName" must be a string AND at least one character ("${cl.name}")`);
        }
        return options.options.customName;
    }
    return baseName;
}
exports.getName = getName;
/**
 * Returns if it is not defined in "schemas"
 * @param cl The Type
 */
function isNotDefined(cl) {
    return typeof cl === 'function' &&
        !isPrimitive(cl) &&
        cl !== Object &&
        cl !== mongoose.Schema.Types.Buffer &&
        util_1.isNullOrUndefined(data_1.schemas.get(getName(cl)));
}
exports.isNotDefined = isNotDefined;
/**
 * Assign "__uniqueID" to a class
 * (used for the decoratorCache)
 * @param cl
 * @returns the initname to be used as identifier
 */
function createUniqueID(cl) {
    if (util_1.isNullOrUndefined(cl.__uniqueID)) {
        cl.__uniqueID = Date.now();
    }
    const initname = `${cl.constructor.name}_${cl.__uniqueID}`;
    if (!data_1.decoratorCache.get(initname)) {
        data_1.decoratorCache.set(initname, { class: cl.constructor, decorators: new Map() });
    }
    return initname;
}
exports.createUniqueID = createUniqueID;
