// inversify-agenda/dist/inversify-agenda.spec.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
require("reflect-metadata");
const inversify_agenda_1 = require("./inversify-agenda");
describe('AgendaTask', () => {
    let TestCommand = class TestCommand {
        execute() {
            return __awaiter(this, void 0, void 0, function* () {
                console.log('DO NOTHING');
            });
        }
    };
    TestCommand = __decorate([
        inversify_agenda_1.task('test', '10 minutes')
    ], TestCommand);
    let TestCommand2 = class TestCommand2 {
        execute() {
            return __awaiter(this, void 0, void 0, function* () {
                console.log('DO NOTHING');
            });
        }
    };
    TestCommand2 = __decorate([
        inversify_agenda_1.task('test2', '20 minutes')
    ], TestCommand2);
    let TestCommand22 = class TestCommand22 {
        execute() {
            return __awaiter(this, void 0, void 0, function* () {
                console.log('DO NOTHING');
            });
        }
    };
    TestCommand22 = __decorate([
        inversify_agenda_1.task('test22', '20 minutes')
    ], TestCommand22);
    it('should invoke decorate class with @inject() add task definition and interval definition', () => {
        expect(inversify_agenda_1.inversifyAgendaTasksConfiguration.tasks.find(t => t.key === 'test')).toEqual({
            key: 'test',
            target: TestCommand
        });
        expect(inversify_agenda_1.inversifyAgendaTasksConfiguration.tasks.find(t => t.key === 'test2')).toEqual({
            key: 'test2',
            target: TestCommand2
        });
        expect(inversify_agenda_1.inversifyAgendaTasksConfiguration.tasks.find(t => t.key === 'test22')).toEqual({
            key: 'test22',
            target: TestCommand22
        });
        expect(inversify_agenda_1.inversifyAgendaTasksConfiguration.intervals['10 minutes']).toEqual([{ key: 'test' }]);
        expect(inversify_agenda_1.inversifyAgendaTasksConfiguration.intervals['20 minutes']).toEqual([{ key: 'test2' }, { key: 'test22' }]);
    });
});
